keps/sig-cluster-lifecycle/0015-kubeadm-join-master.md

```diff
@@ -0,0 +1,431 @@
+  document is referred as “dynamic bootstrap workflow”.
+
+  Nevertheless, kubeadm should support also more “static bootstrap flow”, where a user knows
+  in advance the target layout of the controlplane instances (the number, the name and the IP
+  of master nodes).
+
+### Non-goals
+
+- This proposal doesn't include a solution for etcd cluster management (but nothing in this proposal should
+  prevent to address this in future).
+
+  > At the time of writing, the CoreOS recommended approach for etcd is to run
+  > the etcd cluster outside kubernetes (see discussion in [kubeadm office hours](https://goo.gl/fjyeqo)).
+
+- Nothing in this proposal should prevent users to run master nodes components
+  and etcd on the same machines.
+
+- This proposal doesn't include a solution for API server load balancing (Nothing in this proposal
+  should prevent users from choosing their preferred solution for API server load balancing.
+
+- This proposal doesn't address the ongoing discussion about kubeadm self-hosting; in light of
+  divide and conquer goal stated before, it is not planned to provide initial support
+  for self-hosted clusters (but nothing in this proposal should prevent to address this in future).
+
+- This proposal doesn't address the ongoing discussion about storage of kubeadm TLS assets
+  in secrets; in light of divide and conquer goal stated before, it is not planned to
+  provide initial support for clusters with TLS stored in secrets (but nothing in this
+  proposal should prevent to address this in future).
+
+- Nothing in this proposal should prevent practices that exist today.
+
+### Challenges and Open Questions
+
+- Keep the UX simple.
+
+  - _What are the acceptable trade-offs between the need to have a clean and simple
+    UX and the variety/complexity of possible kubernetes HA deployments?_
+
+- Create a cluster without knowing its final layout
+
+  Supporting a dynamic workflow implies that some information about the cluster are
+  not available at init time, like e.g. the number of master nodes, the IP of
+  master nodes etc. etc.
+
+  - _How to configure a Kubernetes cluster in order to easily adapt to future change
+    of its own controlplane layout like e.g. add a master node, remove a master node?_
+
+  - _What are the "pivotal" cluster settings that must be defined before initialising
+    the cluster?_
+
+  - _How to combine into a single UX support for both static and dynamic bootstrap
+    workflows?_
+
+- Kubeadm limited scope of action
+
+  - Kubeadm binary can execute actions _only_ on the machine where it is running
+    e.g. it is not possible to execute actions on other nodes, to copy files across
+    nodes etc.
+  - During the join workflow, kubeadm can access the cluster _only_ using identities
+    with limited grants, `system:unauthenticated` or `system:node-bootstrapper`.
+
+- Upgradability
+
+  - How to setup an high available cluster in order to simplify the execution
```

Comment:
If we implement  `kubeadm join --master` we MUST support upgrades pretty seamlessly

Reply:
100% agreed! never the less it is a challenge worthy of note here